pym/bob/scm/git.py

```diff
@@ -262,8 +265,22 @@ async def __checkoutTag(self, invoker, fetchCmd, switch):
                 await invoker.checkCommand(["git", "symbolic-ref",
                     "-q", "HEAD"], cwd=self.__dir)
 
-            await invoker.checkCommand(["git", "checkout", "-q", "--no-recurse-submodules",
-                self.__commit if self.__commit else "tags/"+self.__tag], cwd=self.__dir)
+            # test if the configured commit is on the configured branch
+            if self.__branch and self.__commit and self.__useBranchAndCommit:
+                branches = await invoker.checkOutputCommand(["git", "branch", "-a",
+                    "--contains", self.__commit], cwd=self.__dir)
+                if not re.search( r".*/origin/"+re.escape(self.__branch), branches, re.MULTILINE):
+                    raise BuildError ("Commit: " + self.__commit + " is not on branch "
+                        + self.__branch)
+                # if we have both checkout the branch and move to the
+                # commit
+                await self.__checkoutBranch(invoker, fetchCmd, switch)
+
+                await invoker.checkCommand(["git", "reset", "--keep",
+                    "--no-recurse-submodules", self.__commit], cwd=self.__dir)
```

Comment:
I think this is dangerous if some sub-module was added between the commit and the current HEAD of the branch...